### PR TITLE
Fix code actions on Eclipse JDT Language server

### PR DESCRIPTION
**Pull request by beji**
jdtls breaks the spec, which in turn breaks code actions. There is no real solution to this issue so this behaviour needs a workaround.

This was already implemented in eglot but got removed removed when https://github.com/joaotavora/eglot/pull/864 was merged so the corresponding code is reintroduced.

See https://github.com/eclipse/eclipse.jdt.ls/issues/376 for the upstream issue.

### CODE REVIEW

File: eglot.el [hunk 1]
```diff
@@ -3046,6 +3046,18 @@ at point.  With prefix argument, prompt for ACTION-KIND."
...
 
+
+;;; eclipse-jdt breaks the spec which in turn breaks code actions
+;;; This behaviour can't be disabled and needs to be worked around
+;;;
+(defclass eglot-eclipse-jdt (eglot-lsp-server) ()
+  :documentation "Eclipse's Java Development Tools Language Server.")
+
+(cl-defmethod eglot-execute-command
+  ((_server eglot-eclipse-jdt) (_cmd (eql java.apply.workspaceEdit)) arguments)
```
Comment (joaotavora) [inline eglot.el:3057]: You might as well specialize to `server` here, as I don't think many other servers will send `java.apply.workspaceEdit`  (note that it's prefixed by "java").  Then the `eglot-eclipse-jdt` class is not needed.

And the resulting snippet:
```lisp
(cll-defmethod eglot-execute-command
  (_server (_cmd (eql java.apply.workspaceEdit)) arguments)
  "Eclipse JDT breaks spec and replies with edits as arguments."
  (mapc #'eglot--apply-workspace-edit arguments))
```

Become small enough that it _could_ just live in your `.emacs`. But that's arguable , and I'm willing to add these 3 lines of server-specific code to Eglot.

Comment (beji) [reply eglot.el]: That seems to work fine, thanks for your input! Ideally the issue would be fixed (or at least made configurable) in the language server but I don't have high hopes for that after all this time.
I think most users of that language server will most likely want this workaround so it should either be included in eglot or at least documented in the readme file.

**Comment by chvp** [comment]
What's the status on this? As far as I can tell all requested changes have been made?

**Comment by bhankas** [comment]
ping! @joaotavora any chance this can get in?

**Comment by Qkessler** [comment]
Ping again! @joaotavora  Any progress on this? Any blockers we should be aware of or more work that needs to be done and you are in need of a contributor?

Thanks,
Quique.

**Comment by Qkessler** [comment]
Hi @joaotavora, I doesn't quite work with Emacs master. I have had the snippet for a long time in my config, and I'm running into the same issue:
```
java -version
openjdk version "11.0.19" 2023-04-18 LTS
OpenJDK Runtime Environment Corretto-11.0.19.7.1 (build 11.0.19+7-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.19.7.1 (build 11.0.19+7-LTS, mixed mode)
```
Here's the `emacs-repository-version` variable, where `emacs-version` is `30.0.50`.
```
emacs-repository-version is a variable defined in version.el.gz.

Value
"d2548191f116e3ebdf644c3dd281d6e925f51bcf"
```

Thanks for checking!

**Comment by archer-n** [comment]
Code action on the Eclipse JDT language server is available on the emacs master branch. The patch below is no longer needed due to the implementation of :codeAction.resolveSupport .
```elisp
(cll-defmethod eglot-execute-command
  (_server (_cmd (eql java.apply.workspaceEdit)) arguments)
  "Eclipse JDT breaks spec and replies with edits as arguments."
  (mapc #'eglot--apply-workspace-edit arguments))

```

```code
emacs version: GNU Emacs 30.0.50 (build 3, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8) of 2023-07-12 
branch:  emacs master(bc1e7b9b8e681346e407c4dc32272a0bf4decd68)

jdtls version: 1.26.0

```

**Comment by joaotavora** [comment]
Great!! Pity noone told me of this fix before (or maybe they did and I didn't pay  enough attention)

**Comment by LuciusChen** [comment]
`cll-defmethod` should be `cl-defmethod`

**Comment by joaotavora** [comment]
> cll-defmethod should be cl-defmethod

I fixed the comment.  Anyway, the comment says path such a patch is _not_ needed anymore.

**Pull request closed without merging**